activerecord/lib/active_record/reflection.rb

```diff
@@ -804,6 +804,8 @@ def derive_fk_query_constraints(foreign_key)
             MSG
           end
 
+          return foreign_key if primary_query_constraints.include?(foreign_key)
```

Comment:
The expected return of this method is either an array of both pk and fk or an exception. 
https://github.com/rails/rails/pull/50347/files#diff-b1aed4985f41ad87dc1be30472b728a19f1a7354f25ce71bf6ce9a7744a9cd0dR814

This change may break unhidingly. So I consider to rather fix the case in the #foreign_key and add a test case to demonstrate a fix. The foreign key should handle the result and return the right key when dealing with this method.

Reply:
![Uploading Screenshot 2023-12-16 at 12.02.32 AM.png…]()

Reply:
I have prepared a suggestion for you. My suggestion passes your provided test cases reported in the bug https://github.com/rails/rails/issues/50346 

My suggestion is here: https://github.com/rubyrider/rails/pull/2

Feel free if you like to pick it up. :)

Reply:
@rubyrider, I could not find this method being used anywhere else in the code. I don't anticipate any issues, the return type of its caller `foreign_key` seems to adhere to a similar pattern. However, if you've found something that suggests otherwise, could you provide a script to reproduce that scenario?

As for moving the logic to the foreign_key method, I believe it is best placed here as all the behaviour related to the query constraints is encapsulated in this method.

I reviewed the changes you proposed. While they address the current scenario, they don't extend the behavior to provide support for it. What I mean by that is they don't take into account the valid cases where the expected result is an array, which is the case with every other scenario involving query constraints except the current one.

Reply:
so my solution actually fixes the test you provided in your issue. you may check :) Basically the issue what I could see was with the foreign_key. well I am hapoy to correct myself. :)

Reply:
Well, It is not about fixing a test case. In general, it should be about fixing a behaviour or extending it. Query constraints were introduced to make efficient queries using composite keys. You are changing that behaviour by turning the derived key to a single valued attribute again.

I request **Core members** or anyone familiar with the Query Constraints feature to correct me if I'm wrong. Feel free to add any other comments as well.

Let me try to convey my point using the same example. I will make a comparison on the return values for the `foreign_key` method which is going to be used to perform a query on the association and manage stale state.

---
#### Without Query Constraints

`comment.author`

- Expected : author_id
- Actual : author_id

`comment.tenant`

- Expected: tenant_id
- Actual : tenant_id
---
#### With Query Constraints

`comment.author`

- Expected : [tenant_id, author_id]
- Actual : [tenant_id, author_id]

`comment.tenant`

- Expected: tenant_id
- **Actual : [tenant_id, tenant_id]**
- This is the case which needs to be fixed
---

For additional context, When using query constraints expected query to fetch an Author for a comment should be 

`SELECT "authors".* FROM "authors" WHERE "authors"."tenant_id" = ? AND "authors"."id" = ? LIMIT ?` 


I hope it makes sense.

Reply:
@iamradioactive got your point! I am happy to work on this. So is there someone from the core who can contribute a little to  Query Constraints feature?